grin/src/sync.rs

```diff
@@ -49,6 +52,10 @@ pub struct Syncer {
 	last_header_req: Mutex<Instant>,
 	blocks_to_download: Mutex<Vec<Hash>>,
 	blocks_downloading: Mutex<Vec<BlockDownload>>,
+
+	//keep a map of the last header successfully received
+	//from each peerj
+	last_headers_from_peers: Mutex<HashMap<SocketAddr, u64>>,
```

Comment:
I don't think this is enough. Check the comment about the `get_locator` impl below. You could either have a full header clone here or a subset of (height, hash). And while you're at it, you should probably throw in the TD at that block as well.

Reply:
Yep, understood, will change that u64 to a 'BuddyInfo' struct or the like and put all that info in there.